### PR TITLE
Abdu acl

### PR DESCRIPTION
**Pull request by ghanemabdo**
work-in-progress. Added permissions.js that enables apps to programatically modify acls of a resource. Still needs a lot of testing for all possible use cases.

### CODE REVIEW

File: lib/permissions.js [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+    return new Promise(function (resolve, reject) {
+      // reading the existing acls from the server and merging them with the new acls for each agent
+      mergeAcls(url, agents, permissions).then(
+        function (mergedPermissions) {
+          // having the merged permissions, create the RDF string representing the auth rules
+          // for the list of agents
+          var aclString = createAclRules(url, agents, permissions)
+          aclString = appendOwnerRule(aclString, url)
+          // write the new acl resource via put http method to replace any existing one
+          webClient.put(url + ',acl', aclString).then(
```
Comment (nicola) [inline lib/permissions.js]: You are not guaranteed that the ACL will have this name `,acl` before doing this put, as of today, you need to do a HEAD request first and get the Acl uri from the link header :)

Comment (ghanemabdo) [inline lib/permissions.js]: Right. @dmitrizagidulin just pointed me to this. I guess we need to find a consistent way to identify how the naming of acl resources go on the server or at least the pattern (specs) that shows how the naming should be.

Comment (deiu) [inline lib/permissions.js]: This may help https://github.com/solid/solid-spec/blob/master/authz-solid-wac.md

**Comment by dmitrizagidulin** [comment]
@deiu @nicola - I went over the PR with Abdo on Skype this morning. 

I'm going to merge this code to serve as a starting baseline of our ACL functionality in solid.js, and work with Abdo & you guys to refactor it to match our ACL specs. (Which we need to write/clarify). And will also be pulling in functionality from Tim's [tabulator ACL code](https://github.com/linkeddata/tabulator/blob/gh-pages/js/panes/common/), as well.

**Comment by deiu** [comment]
I'd like to take a look at it first.

**Comment by dmitrizagidulin** [comment]
Sure thing.

**Comment by deiu** [comment]
@ghanemabdo I have created a new branch called `permissions`. Can you please send your PR to that branch, so we can easily manage work in progress for permissions? I'm hesitant to merge this PR into master right away. The `permissions` branch gives us more flexibility.

**Comment by ghanemabdo** [comment]
Pushed to permissions branch instead of master.

**Comment by dmitrizagidulin** [comment]
@ghanemabdo Perfect, thanks! Closing this PR, will work on that one.

**Pull request closed without merging**